Cloud Application Manager/Automating Deployments/running-code-outside-an-instance.md

```diff
@@ -0,0 +1,99 @@
+![publish box](../../images/cloud-application-manager/task-variables-05.png)
+* Enter the details of the box as you want them to appear on the catalog. Specify the File variables you want to convert into Task variables.
+![publish details](../../images/cloud-application-manager/task-variables-06.png)
+* Your box will undergo a verification process and, if approved, will eventually be published and appear in the catalog. When you access it, a green checkmark will appear close to the version you published.
+![published version](../../images/cloud-application-manager/task-variables-07.png)
+* After that, the script box can be deployed with a deployment policy as you would normally do. The task variables are not visible during deployment and cannot be altered. Please note that only the versioned, published box will run the scripts as task variables. If you try to run the box draft, it will be deployed normally, but the code in the file variables will not be executed.
+![new instance from published version](../../images/cloud-application-manager/task-variables-08.png)
+* Once you click Deploy, the instance will be created and you will be transferred to the Instance Activity page. Each deploy task variable will appear on its own activity step where the output from the script will be printed. Once all task variables have completed successfully, a new VM will be provisioned in the provider and all event scripts will run as usual.
+![deployment of instance with task variables](../../images/cloud-application-manager/task-variables-09.png)
+* If the script sets a variable value during the task variable execution, those values will be accessible by the event scripts. Note that the task variables are not accessible in the [Lifecycle Editor](../Core Concepts/lifecycle-editor.md).
+![lifecycle editor](../../images/cloud-application-manager/task-variables-10.png)
+* When you terminate the instance, task variables for the *remove* event will be run *after* the instance has been stopped and deleted from the provider.
+
+### Script failures
+
+The exit code of the task variables affects the state of the instance as the event scripts do: any value different from 0 will put the instance in *unavailable* state. If this occurs during the deploy operation, no VM will be provisioned and the **Retry Deploy** action will be available. If this occurs during the terminate operation, the VM will be deleted, but the instance will appear as *unavailable*. At that point you may either **Force Terminate** the instance, triggering the task variables, or **Delete** it, removing the instance from the workspace without running the task variables again.
+
+![failure on task variable](../../images/cloud-application-manager/task-variables-11.png)
+
+Task variables may also fail if the script does not terminate in a timely fashion. If that occurs, the script will be forcibly killed and the message *Task variable took too long, terminating...* will appear in the log.
```

Comment:
What is the "timely fashion"? Do we have a specific timeout set up? Please, specify it

Reply:
I added a 30 second limit plus a 10 extra seconds for a clean shutdown if the first timeout expires. After that, a hard shutdown is triggered. I picked the numbers a bit out of the blue and left this a bit unclear to trigger this discussion. I'll list those numbers in t he KB if they make sense.